### PR TITLE
[wmst/xyz] Use the tile download manager to fetch tiles

### PR DESCRIPTION
**Pull request by nirvn**
## Description

@wonder-sk , @nyalldawson , how does that look?

### CODE REVIEW

**Comment by nirvn** [comment]
Small note on attributes and headers, ATM, I've merely saved the ones needed by the WMS provider, thinking it'd be faster than looping through all possible attribute and header values via QMetaEnum::fromType(...).

**Comment by nirvn** [comment]
OK, this is *very* cool, XYZ drawing feels a lot more responsive, like, by _a lot_.

However, if I zoom in and out a lot and settle on a part of the globe I haven't fetched yet, shutting down QGIS freezes the app. I'm pretty sure this PR causes the regression. 

@wonder-sk , clues as to what might be happening here? Are the canceled draw() not handled properly? That's one of the few differences I'm seeing when looking at the vector tile rendering stuff (i.e. the QgsWmsTiledImageDownloadHandler's handling of its mEventLoop)

**Comment by nirvn** [comment]
It's stuck here:

```
void QgsApplication::exitQgis()
{
  // make sure all threads are done before exiting
  QThreadPool::globalInstance()->waitForDone();
  ...

```

**Comment by nirvn** [comment]
OK, found out the problem, fixed. It is a huge boost in UX, nice.

File: src/core/qgstiledownloadmanager.cpp [hunk 1]
```diff
@@ -124,7 +124,17 @@ void QgsTileDownloadManagerReplyWorkerObject::replyFinished()
       data = reply->readAll();
   }
 
-  emit finished( data, reply->error(), reply->errorString() );
+  QMap<QNetworkRequest::Attribute, QVariant> attributes;
+  attributes.insert( QNetworkRequest::SourceIsFromCacheAttribute, reply->attribute( QNetworkRequest::SourceIsFromCacheAttribute ) );
+  attributes.insert( QNetworkRequest::SourceIsFromCacheAttribute, reply->attribute( QNetworkRequest::RedirectionTargetAttribute ) );
+  attributes.insert( QNetworkRequest::SourceIsFromCacheAttribute, reply->attribute( QNetworkRequest::HttpStatusCodeAttribute ) );
+  attributes.insert( QNetworkRequest::SourceIsFromCacheAttribute, reply->attribute( QNetworkRequest::HttpReasonPhraseAttribute ) );
+  attributes.insert( QNetworkRequest::SourceIsFromCacheAttribute, reply->attribute( QNetworkRequest::HttpReasonPhraseAttribute ) );
```
Comment (wonder-sk) [inline src/core/qgstiledownloadmanager.cpp:132]: this does not seem right, or am I missing something?

File: src/core/qgstiledownloadmanager.h [hunk 2]
```diff
@@ -53,11 +53,18 @@ class CORE_EXPORT QgsTileDownloadManagerReply : public QObject
...
     //! Returns binary data returned in the reply (only valid when already finished)
     QByteArray data() const { return mData; }
+    //! Returns the reply URL
+    QUrl url() const { return mUrl; }
+    //! Returns the attribute associated with the \a code
+    QVariant attribute( QNetworkRequest::Attribute code );
+    //! Returns the value of the known header \a header.
+    QVariant header( QNetworkRequest::KnownHeaders header );
+    //! Returns a list of raw header pairs
+    const QList<QNetworkReply::RawHeaderPair> rawHeaderPairs() const { return mRawHeaderPairs; }
```
Comment (wonder-sk) [inline src/core/qgstiledownloadmanager.h:63]: Just curious - were these changes necessary for something, or have you added them just in case? They do not seem to be used anywhere in the client code...

**Review by wonder-sk — COMMENTED** [review]
Thanks for looking into this! The WMS provider changes look sensible, just some questions about the tile download manager changes...

File: src/core/qgstiledownloadmanager.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/qgstiledownloadmanager.cpp:132]: @wonder-sk , indeed there's a duplicate here.

File: src/core/qgstiledownloadmanager.h (hunk 2 above)
Comment (nirvn) [reply src/core/qgstiledownloadmanager.h:63]: Yeah, QgsWmsTiledImageDownloadHandler::tileReplyFinished() makes use of all these.

Comment (wonder-sk) [reply src/core/qgstiledownloadmanager.h:63]: Ahh right, good point, it just didn't show up in the diff and I got confused :-)

File: src/core/qgstiledownloadmanager.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/core/qgstiledownloadmanager.cpp:132]: Not just the duplicate, but also `SourceIsFromCacheAttribute` is inserted multiple times...

It's a pity there doesn't seem to be a way to get all attributes and one needs to go one by one.

**Comment by nirvn** [comment]
@wonder-sk , comment addressed.

**Review by wonder-sk — APPROVED** [review]
